src/collapse/xdebug.rs

```diff
@@ -0,0 +1,367 @@
+use super::Collapse;
+use hashbrown::HashMap;
+use hashbrown::hash_map::RawEntryMut;
+use log::warn;
+use std::fmt;
+use std::hash::{BuildHasher, Hash, Hasher};
+use std::io::prelude::*;
+use std::io::{self, Write};
+use std::rc::Rc;
+// Ideas: str_stack and colosseum to keep allocations closer together or something.
+
+const SCALE_FACTOR: f32 = 1_000_000.0;
+static CALLS: &[&str] = &["require", "require_once", "include", "include_once"];
+
+const TRACE_START: &str = "TRACE START";
+const TRACE_END: &str = "TRACE END";
+
+/// Options for the Xdebug collapser
+#[derive(Clone, Debug, Default)]
+pub struct Options;
+
+/// A unique key for an interned string.
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+struct Str(usize);
+
+#[derive(Clone, Copy, PartialEq, Eq, Hash)]
+enum Call {
+    WithPath(Str, Str),
+    WithoutPath(Str),
+}
+
+enum Interned<T> {
+    Old(T),
+    New(T),
+}
+
+#[derive(Default)]
+struct CallStack {
+    strings: HashMap<Rc<str>, usize>,
+    interned_string: Vec<Rc<str>>,
+
+    calls: HashMap<Call, usize>,
+    interned: Vec<Call>,
+
+    stack: Vec<usize>,
+}
+
+struct Frames<'a> {
+    calls: &'a CallStack,
+    stack: &'a [usize],
+}
+
+/// Iterate over tab separated fields.
+///
+/// This is essentially the same as `str.split('\t').filter(|slice| slice.len() > 0)` but seemingly
+/// faster. This probably because we know that the character in question is ascii, while the
+/// standard library `StrSearcher` for `char` is not optimized around this case.
+struct Fields<'a> {
+    line: &'a str,
+}
+
+/// The Folder struct
+pub struct Folder;
+
+impl From<Options> for Folder {
+    fn from(_: Options) -> Self {
+        Self
+    }
+}
+
+impl Collapse for Folder {
+    fn is_applicable(&mut self, input: &str) -> Option<bool> {
+        let mut input = input.as_bytes();
+        let mut line = String::new();
+        loop {
+            line.clear();
+
+            if let Ok(n) = input.read_line(&mut line) {
+                if n == 0 {
+                    break;
+                } else {
+                    return Some(false);
+                }
+            }
+
+            if line.starts_with(TRACE_START) {
```

Comment:
If you know that an xdebug file always starts with `TRACE_START` on the very first line, then this check should probably test specifically for that, and return `Some(false)` otherwise.